Bolt V3

lutovich opened the pull request:
Added new message types and logic for handling them. New metadata like `tx_timeout` and `tx_metadata` are not used yet.

Renamed `NettyConnection` to `DirectConnection` and `Bookmark` to `Bookmarks`.

lutovich committed:
Messaging part of Bolt V3

Added new messages and ability to write them into a network channel.
Introduced Bolt V3 abstraction that uses new messages to execute
execute queries, begin/commit/rollback transactions.

Code does not yet use new metadata.

lutovich committed:
Rename NettyConnection to DirectConnection

lutovich committed:
Rename Bookmark to Bookmarks

This is a better name because class holds a collection of bookmarks,
not just a single one.

lutovich committed:
Fix ErrorIT to use correct message format

It always used V1 message format before.

lutovich committed:
Fix handing of response metadata in Bolt V3

Metadata keys are different in previous versions of the protocol.
This commit makes it possible for keys to be defined by each Bolt
protocol implementation.

Also fixed couple unit tests and added license headers.

zhenlineo commented on driver/src/main/java/org/neo4j/driver/internal/handlers/PullAllResponseHandler.java:
```diff
@@ -66,10 +67,11 @@
     private CompletableFuture<Throwable> failureFuture;
 
-    public PullAllResponseHandler( Statement statement, RunResponseHandler runResponseHandler, Connection connection )
+    public PullAllResponseHandler( Statement statement, RunResponseHandler runResponseHandler, String resultConsumedAfterMetadataKey, Connection connection )
```
I can see why you have this key here, 
but how about override this class and then have two implementation of `extractResultSummary` method? Versioning response handler sound a reasonable thing.

```
public ResultSummary extractResultSummary( Map<String,Value> metadata ) {
  long resultAvailableAfter = runResponseHandler.resultAvailableAfter();
  return MetadataUtil.extractSummary( statement, connection, resultAvailableAfter, metadata, T_LAST_KEY );
  // or RESULT_COMSUMED_AFTER_KEY for v1 and v2
}
```

Similarly for `RunResponseHandler`, we can override `onSuccess` or `onMeta`?

lutovich replied:
👍 I'll give this a try. Might especially be valuable if we add another response metadata field

lutovich replied:
Made the metadata extraction utility versioned instead of passing string keys around. Creating classes felt a little to much - 3 new classes would need to be created just to override 2 string keys

zhenlineo commented on driver/src/main/java/org/neo4j/driver/internal/messaging/request/TransactionStartingMessage.java:
```diff
@@ -0,0 +1,70 @@
+
+    final Map<String,Value> metadata;
+
+    TransactionStartingMessage( Bookmarks bookmarks, Duration txTimeout, Map<String,Value> txMetadata )
```
aha, smart interface to take all transaction related metadata

zhenlineo commented on driver/src/test/java/org/neo4j/driver/internal/handlers/HelloResponseHandlerTest.java:
```diff
@@ -0,0 +1,135 @@
+        ChannelPromise channelPromise = channel.newPromise();
+        HelloResponseHandler handler = new HelloResponseHandler( channelPromise );
+
+        assertThrows( IllegalArgumentException.class, () -> handler.onSuccess( singletonMap( "server", value( "WrongServerVersion" ) ) ) );
```
Is it still required to have the server version parsed rather than simply record it as a simple message?

lutovich replied:
Server version should not be needed for version comparisons in Bolt V3. It is just easier to still keep it as `ServerVersion` object for older protocol versions

lutovich committed:
Version metadata extraction utility

Instead of passing string metadata keys around. This is needed because
some metadata keys have changed in Bolt V3 compared to Bolt V1.
